core/src/test/java/org/apache/iceberg/TestRowLineageMetadata.java

```diff
@@ -0,0 +1,339 @@
+  }
+
+  @ParameterizedTest
+  @FieldSource("org.apache.iceberg.TestHelpers#ALL_VERSIONS")
+  public void testDeletes(int formatVersion) {
+    assumeTrue(formatVersion >= TableMetadata.MIN_FORMAT_VERSION_ROW_LINEAGE);
+
+    TestTables.TestTable table =
+        TestTables.create(
+            tableDir, "test", TEST_SCHEMA, PartitionSpec.unpartitioned(), formatVersion);
+    TableMetadata base = table.ops().current();
+    table.ops().commit(base, TableMetadata.buildFrom(base).enableRowLineage().build());
+
+    assertThat(table.ops().current().rowLineage()).isTrue();
+    assertThat(table.ops().current().lastRowId()).isEqualTo(0L);
+
+    DataFile file = fileWithRows(30);
+
+    table.newAppend().appendFile(file).commit();
+
+    assertThat(table.ops().current().rowLineage()).isTrue();
+    assertThat(table.currentSnapshot().firstRowId()).isEqualTo(0);
+    assertThat(table.ops().current().lastRowId()).isEqualTo(30);
+
+    table.newDelete().deleteFile(file).commit();
+
+    // Deleting a file should create a new snapshot which should inherit last-row-id from the
+    // previous metadata and not
+    // change last-row-id for this metadata.
+    assertThat(table.ops().current().rowLineage()).isTrue();
+    assertThat(table.currentSnapshot().firstRowId()).isEqualTo(30);
+    assertThat(table.currentSnapshot().addedRows()).isEqualTo(0);
+    assertThat(table.ops().current().lastRowId()).isEqualTo(30);
+  }
+
+  @ParameterizedTest
+  @FieldSource("org.apache.iceberg.TestHelpers#ALL_VERSIONS")
+  public void testReplace(int formatVersion) {
+    assumeTrue(formatVersion >= TableMetadata.MIN_FORMAT_VERSION_ROW_LINEAGE);
+
+    TestTables.TestTable table =
+        TestTables.create(
+            tableDir, "test", TEST_SCHEMA, PartitionSpec.unpartitioned(), formatVersion);
+    TableMetadata base = table.ops().current();
+
+    table.ops().commit(base, TableMetadata.buildFrom(base).enableRowLineage().build());
+
+    assertThat(table.ops().current().rowLineage()).isTrue();
+    assertThat(table.ops().current().lastRowId()).isEqualTo(0L);
+
+    DataFile filePart1 = fileWithRows(30);
+    DataFile filePart2 = fileWithRows(30);
+    DataFile fileCompacted = fileWithRows(60);
+
+    table.newAppend().appendFile(filePart1).appendFile(filePart2).commit();
+
+    assertThat(table.ops().current().rowLineage()).isTrue();
+    assertThat(table.currentSnapshot().firstRowId()).isEqualTo(0);
+    assertThat(table.ops().current().lastRowId()).isEqualTo(60);
+
+    table.newRewrite().deleteFile(filePart1).deleteFile(filePart2).addFile(fileCompacted).commit();
+
+    // Rewrites are currently just treated as appends. In the future we could treat these as no-ops
+    assertThat(table.ops().current().rowLineage()).isTrue();
+    assertThat(table.currentSnapshot().firstRowId()).isEqualTo(60);
+    assertThat(table.ops().current().lastRowId()).isEqualTo(120);
+  }
+
+  @ParameterizedTest
+  @FieldSource("org.apache.iceberg.TestHelpers#ALL_VERSIONS")
```

Comment:
it looks like all of the methods are being executed with the same parameters. In that case it makes sense to do the parameterization at the class level. A simple example test is `TestFastAppend` where the class is annotated with `@ExtendWith(ParameterizedTestExtension.class)` and tests with `@TestTemplate`

Reply:
This is something I wanted to ask you about, I am not a fan of the ExtendWith ParameterizedTestExtension, it feels like we are working against the framework we choose to use. I understand it was important for backwards compatability with our existing tests but do we really want to keep be using it in the future? 

When I was working on some other tests for the schema compatibility checks, having the ExtendsWith basically blocks us from using any other Junit annotations for parameter passing. For example passing two sets of parameters. With this in mind I was wondering if it makes sense to not rely on the ParameteriazedTestExtension in the future.

Reply:
It is true that the parameterization at the class level makes it impossible to use other annotations on a single test method, but most of the time we would still parameterize at the class level and skip running a particular test method for a certain parameter. 
I don't have a strong opinion here and if you think these methods will eventually all have separate test parameters, then parameterizing at the method level is probably better. 
However, currently it rather looks like the only parameters across all tests would be the format versions.

Also btw the `ParameterizedTestExtension` was not only due to backwards compatibility during the migration process, but also as an intermediate solution until JUnit5 fully supports parameterization at the class level

Reply:
Does Junit plan to support class level parameterization? I would be fully on board with it in that case, I just don't want to be stuck with the extension indefinitely.

Reply:
according to https://github.com/junit-team/junit5/issues/878 it seems that this will eventually be supported in JUnit5

Reply:
Ok i'll make the switch

Reply:
That Issue has been open for 7 years :|

Reply:
> That Issue has been open for 7 years :|

yeah I know :( but according to recent comments it seems that they are planning to release something in JUnit 5.13